=============================
vcffuse
=============================

.. image:: https://badge.fury.io/py/vcffuse.png
    :target: http://badge.fury.io/py/vcffuse

.. image:: https://travis-ci.org/szilvajuhos/vcffuse.png?branch=master
    :target: https://travis-ci.org/szilvajuhos/vcffuse

Visualising gene fusions in VCFs
================================

The main purpose is to visualise structural variants in a similar way
like in this_ article.

.. _this: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5728033/figure/onco12246-fig-0006/?report=objectonly


Features
--------

Presently it is restricted to file format generated by Manta_ and
annotated by snpEff_ . Goal is to follow structural variants as they are
described in the `VCF Specification`_

.. _Manta: https://github.com/Illumina/manta
.. _snpEff: https://pcingola.github.io/SnpEff/
.. _VCF Specification: https://samtools.github.io/hts-specs/VCFv4.2.pdf

* TODO
    - generally refactor code, actual steps below
    - add test cases
    - refactor all the functions that are ad-hoc added to code
    - check tandem repeats https://github.com/szilvajuhos/vcffuse/blob/master/vcffuse/getFusionFromVCF.py#L198
    - DRY in
        - fuse_deletion()
        - fuse_tandem_genes()
        - fuse_translocations()

* Usage:

It is not a module yet, add a conda environment stored in ``environment.yml`` and try

``$ python vcffuse/getFusionFromVCF.py``

